### ISSUE TITLE
glfw/input.c v0.15.0 fails to compile with time errors

### ISSUE THREAD

**Issue by scopatz** [kovidgoyal/kitty#2172]
With v0.15.0, I am seeing the following compilation errors ([full log here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=98482)):

```
[K[7/59] Compiling [x11] glfw/input.c ...
[K[8/59] Compiling kitty/parser.c ... done
Compiling [x11] glfw/input.c ...
$BUILD_PREFIX/bin/x86_64-conda_cos6-linux-gnu-cc -MMD -DNDEBUG -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem $PREFIX/include -D_GLFW_X11 -D_GLFW_BUILD_DLL -Wextra -Wfloat-conversion -Wno-missing-field-initializers -Wall -Wstrict-prototypes -std=c11 -pedantic-errors -Werror -O3 -fwrapv -fstack-protector-strong -pipe -fvisibility=hidden -fPIC -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/kitty-0.15.0 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -I $PREFIX/include/freetype2 -I $PREFIX/include/harfbuzz -DTIME_WITH_SYS_TIME=1 -flto -fpic -pthread -I$PREFIX/include -I$PREFIX/include -I$PREFIX/include -I$PREFIX/include -I$PREFIX/include -I$PREFIX/include -I$PREFIX/include -I$PREFIX/include/dbus-1.0 -I$PREFIX/lib/dbus-1.0/include -c glfw/input.c -o build/glfw-x11-input.c.o
In file included from glfw/internal.h:30:0,
                 from glfw/input.c:30:
glfw/../kitty/monotonic.h:19:49: error: 'struct timespec' declared inside parameter list will not be visible outside of this definition or declaration [-Werror]
 static inline monotonic_t calc_nano_time(struct timespec time) {
                                                 ^~~~~~~~
glfw/../kitty/monotonic.h:19:58: error: parameter 1 ('time') has incomplete type
 static inline monotonic_t calc_nano_time(struct timespec time) {
                                                          ^~~~
glfw/../kitty/monotonic.h:19:27: error: function declaration isn't a prototype [-Werror=strict-prototypes]
 static inline monotonic_t calc_nano_time(struct timespec time) {
                           ^~~~~~~~~~~~~~
glfw/../kitty/monotonic.h: In function 'calc_nano_time':
glfw/../kitty/monotonic.h:19:58: error: unused parameter 'time' [-Werror=unused-parameter]
 static inline monotonic_t calc_nano_time(struct timespec time) {
                                                          ^~~~
glfw/../kitty/monotonic.h: At top level:
glfw/../kitty/monotonic.h:28:31: error: return type is an incomplete type
 static inline struct timespec calc_time(monotonic_t nsec) {
                               ^~~~~~~~~
glfw/../kitty/monotonic.h: In function 'calc_time':
glfw/../kitty/monotonic.h:29:21: error: storage size of 'result' isn't known
     struct timespec result;
                     ^~~~~~
glfw/../kitty/monotonic.h:32:12: error: 'return' with a value, in function returning void
     return result;
            ^~~~~~
glfw/../kitty/monotonic.h:28:31: note: declared here
 static inline struct timespec calc_time(monotonic_t nsec) {
                               ^~~~~~~~~
glfw/../kitty/monotonic.h:29:21: error: unused variable 'result' [-Werror=unused-variable]
     struct timespec result;
                     ^~~~~~
glfw/../kitty/monotonic.h: In function 'monotonic_':
glfw/../kitty/monotonic.h:72:12: error: variable 'ts' has initializer but incomplete type
     struct timespec ts = {0};
            ^~~~~~~~
glfw/../kitty/monotonic.h:72:27: error: excess elements in struct initializer
     struct timespec ts = {0};
                           ^
glfw/../kitty/monotonic.h:72:27: note: (near initialization for 'ts')
glfw/../kitty/monotonic.h:72:21: error: storage size of 'ts' isn't known
     struct timespec ts = {0};
                     ^~
glfw/../kitty/monotonic.h:78:5: error: implicit declaration of function 'clock_gettime'; did you mean 'calc_time'? [-Wimplicit-function-declaration]
     clock_gettime(CLOCK_MONOTONIC, &ts);
     ^~~~~~~~~~~~~
     calc_time
glfw/../kitty/monotonic.h:78:19: error: 'CLOCK_MONOTONIC' undeclared (first use in this function)
     clock_gettime(CLOCK_MONOTONIC, &ts);
                   ^~~~~~~~~~~~~~~
glfw/../kitty/monotonic.h:78:19: note: each undeclared identifier is reported only once for each function it appears in
glfw/../kitty/monotonic.h:72:21: error: unused variable 'ts' [-Werror=unused-variable]
     struct timespec ts = {0};
                     ^~
In file included from glfw/internal.h:30:0,
                 from glfw/input.c:30:
glfw/../kitty/monotonic.h:81:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1: all warnings being treated as errors
```

The [build script](https://github.com/conda-forge/kitty-feedstock/blob/3742cc15ededbfd9119368043941e4861bd9bedb/recipe/build.sh) is:

```sh
#!/bin/bash
set -ex

export PKG_CONFIG_PATH="${PREFIX}/lib/pkgconfig:/usr/lib64/pkgconfig:/usr/lib/x86_64-linux-gnu/pkgconfig:${PKG_CONFIG_PATH}"
export CFLAGS="${CFLAGS} -I ${PREFIX}/include/freetype2 -I ${PREFIX}/include/harfbuzz"
$PYTHON setup.py linux-package --prefix=$PREFIX --update-check-interval=0

# add bash completions
mkdir -p "${PREFIX}/share/bash-completion/completions"
"${PREFIX}/bin/kitty" + complete setup bash > "${PREFIX}/share/bash-completion/completions/kitty"
```

[Some places online](https://stackoverflow.com/questions/26769129/trying-to-use-clock-gettime-but-getting-plenty-of-undeclared-errors-from-ti) indicate that this do to a missing define, but when I added `-D_POSIX_C_SOURCE=199309L` to CFLAGS, I actually saw that `_POSIX_C_SOURCE 200809L`. This is a much more recent version. [Full logs here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=98734).

[The Python bug tracker](https://bugs.python.org/issue25853) then indicated that I should set `TIME_WITH_SYS_TIME 1`, so I tried setting this in the `CFLAGS` instead. But this just returned me to the original error.

**Comment by kovidgoyal** [comment]
What compiler/c library are you using? From your log it looks like you might be using anaconda. Figure out what is needed to get time.h to define clock_gettime() on whatever system you are using.

**Comment by scopatz** [comment]
this is gcc on centos6 with the conda-forge compilers. It is all in the  logs linked there

**Comment by Luflosi** [comment]
@kovidgoyal why did that change break stuff? `clock_gettime()` was used before and that worked.

**Comment by kovidgoyal** [comment]
@scopatz sorry I have no experience with that toolchain. As per the man page of clock_gettime you need to define _POSIX_C_SOURCE. The kitty build system does that already as you discovered. Why it does not work on your system, I haven't a clue. Something somewhere is broken. I'm afraid I dont have much else to tell you.

@Luflosi the change you made created monotonic.h which gets included in a ton of places. Which means it is no longer possible to guarantee time.h is included for the first time by monotonic.h. Which means that the macros you need for clock_gettime have to be defined as compiler arguments to ensure they are present the first time time.h is included.

**Comment by Luflosi** [comment]
Would including `monotonic.h` above all other includes everywhere help?

**Comment by kovidgoyal** [comment]
Not really since the needed macro is now defined as a compiler flag. As
far as this bug report goes, the problem is the OPs compiler/libc is not
respecting the macro, so there's not anything that we can do about it

**Comment by scopatz** [comment]
As I mentioned in the original post, `_POSIX_C_SOURCE` is defined to `200809L` which makes me think that kitty is broken on any releatively modern GCC toolchain, and that this isn't an issue specific to me. 

> the problem is the OPs compiler/libc is not respecting the macro, so there's not anything that we can do about it

This is not true. I think it is a shame that you close all my issues so quickly without really considering. You understand your code base far better than I could and could help me work through any compiler issues.

**Comment by Luflosi** [comment]
Don't worry too much about this issue being closed, you can still continue this discussion and hopefully find a solution.

**Comment by kovidgoyal** [comment]
On Mon, Dec 02, 2019 at 01:06:50PM -0800, Anthony Scopatz wrote:
> As I mentioned in the original post, `_POSIX_C_SOURCE` is defined to `200809L` which makes me think that kitty is broken on any releatively modern GCC toolchain, and that this isn't an issue specific to me. 

I build kitty on my machine using gcc version 9.2.0 and in fact to get
kitty to build with musl libc I added _POSIX_C_SOURCE=200809 to the
kitty build system a while ago. 
https://github.com/kovidgoyal/kitty/commit/2cee3a88094d2421e74738e59de97abd3302a20f

> 
> > the problem is the OPs compiler/libc is not respecting the macro, so there's not anything that we can do about it
> 
> This is not true. I think it is a shame that you close all my issues so quickly without really considering. You understand your code base far better than I could and could help me work through any compiler issues.

I do understand my code base better than you could, and therefore I know
the issue is with your compiler. You dont need to believe me, read the
man page of clock_gettime().

man 3 clock_gettime

**Comment by scopatz** [comment]
> Don't worry too much about this issue being closed, you can still continue this discussion and hopefully find a solution.

Yeah, if it was just this issue, I wouldn't be so worried. Rather I see this as a pattern of not-very-community oriented behavior. 

I really want to use kitty (and tell other people to se it too!), but my experiances here have been pretty bad. I think I'll just stop using it, unfortunately, rather than trying to help fix issues and getting shot down. 

I am still open to helping out if the kitty team is open to help.

**Comment by Luflosi** [comment]
@scopatz there is a chance that your issue was fixed in 7bf0afa621ccbce8f3a078f7e921e5f5a456257b. Can you test that without a new release?
